File: process_ships.js
const Jimp = require("jimp");
const path = require("path");

async function processImage(filename, targetColor, tolerance) {
  const activePath = path.join(__dirname, "frontend/public/images", filename);
  console.log(`Processing ${activePath}...`);

  try {
    const image = await Jimp.read(activePath);
    const targetR = targetColor[0];
    const targetG = targetColor[1];
    const targetB = targetColor[2];

    image.scan(
      0,
      0,
      image.bitmap.width,
      image.bitmap.height,
      function (x, y, idx) {
        const r = this.bitmap.data[idx + 0];
        const g = this.bitmap.data[idx + 1];
        const b = this.bitmap.data[idx + 2];

        // Check for white background (or close to white)
        // If target is white (255,255,255), we check if r,g,b are all close to 255
        // Or simply check if they are > threshold

        let isMatch = false;

        if (targetR === 255 && targetG === 255 && targetB === 255) {
          // For white, we can be more aggressive: if it's very bright, kill it
          if (r > 240 && g > 240 && b > 240) {
            isMatch = true;
          }
        } else {
          if (
            Math.abs(r - targetR) <= tolerance &&
            Math.abs(g - targetG) <= tolerance &&
            Math.abs(b - targetB) <= tolerance
          ) {
            isMatch = true;
          }
        }

        if (isMatch) {
          this.bitmap.data[idx + 3] = 0; // Set Alpha to 0
        }
      }
    );

    await image.writeAsync(activePath);
    console.log(`Success: ${filename}`);
  } catch (err) {
    console.error(`Error processing ${filename}:`, err);
  }
}

async function main() {
  await processImage("ships/y-wing.png", [255, 255, 255], 30);
}

main();

File: process_ships.py

from PIL import Image
import numpy as np

def make_transparent(image_path, target_bg_color, tolerance=30):
    print(f"Processing {image_path}...")
    try:
        img = Image.open(image_path).convert("RGBA")
        data = np.array(img)

        r_t, g_t, b_t = target_bg_color
        r, g, b, a = data[:, :, 0], data[:, :, 1], data[:, :, 2], data[:, :, 3]

        # Improved mask logic: check for "whiteness" generally
        # If r,g,b are all > 240, it's white.
        mask = (r > 240) & (g > 240) & (b > 240)
        
        # Also apply the specific target color tolerance just in case
        # mask = (np.abs(r - r_t) < tolerance) & (np.abs(g - g_t) < tolerance) & (np.abs(b - b_t) < tolerance)

        data[mask, 3] = 0
        
        new_img = Image.fromarray(data)
        new_img.save(image_path)
        print(f"Saved transparent version to {image_path}")
    except Exception as e:
        print(f"Error: {e}")

make_transparent(
    "frontend/public/images/ships/y-wing.png",
    (255, 255, 255),
    tolerance=40
)
